### PR TITLE
Opaaddon

### PR DESCRIPTION
**Pull request by parkand1**
Confirmed that OPA addon is running and works - will include examples and documentation once reviewers approve

### CODE REVIEW

**Review by kcoleman731 — COMMENTED** [review]
@parkand1 can you add docs for the addon?

File: lib/addons/OPA/index.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { ClusterAddOn, ClusterInfo } from "../../stacks/cluster-types";
+
+export class OpaAddOn implements OpaAddOn {
```
Comment (shapirov103) [inline lib/addons/OPA/index.ts:3]: If it is gatekeeper that we install, a better name would be `OpaGatekeeperAddOn` (based on https://github.com/open-policy-agent/gatekeeper#how-is-gatekeeper-different-from-opa)

File: lib/addons/OPA/index.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    deploy(clusterInfo: ClusterInfo): void {
```
Comment (shapirov103) [inline lib/addons/OPA/index.ts:5]: Please take a look at the supported values.yml for the chart and see if any of those values make sense to be passed explicitly (e.g. namespace, service type). If not, then all values should be allowed similar to what we did in for example Nginx addon (see NginxAddOnProps.values) so that customer could tune it, e.g. setting whether deny attempts are logged, etc.

File: docs/addons/opa-gatekeeper.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+OPA Gatekeeper is an open-source project that provides a first-class integration between OPA and Kubernetes. What Gatekeeper adds is an extensible parameterized policy library that includes native Kubernetes CRD's for instantiating and extending the OPA policy library. Gatekeeper also provides audit functionality as well. The diagram below shows how Gatekeeper interacts with the Kube API Server.
+
+![opa](https://d33wubrfki0l68.cloudfront.net/a5ed0c27ff2dda6abb18b9bc960f2ad4120d937a/a5939/docs/latest/images/kubernetes-admission-flow.png))
+
+In the context of a Shared Services Platform running on Amazon EKS, platform teams and administrators need a way of being able to set policies to adhere to governance and security requirements for all workloads and teams working on the same cluster. Examples of standard use cases for using policies via OPA Gatekeeper are shown below:
+
+- Which users can access which resources.
+- Which subnets egress traffic is allowed to.
+- Which clusters a workload must be deployed to.
```
Comment (shapirov103) [inline docs/addons/opa-gatekeeper.md:13]: Is this applicable to SSP? If yes, then how?

File: docs/addons/opa-gatekeeper.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+- Which registries binaries can be downloaded from.
+- Which OS capabilities a container can execute with.
+- Which times of day the system can be accessed at.
+
+## How does Gatekeeper work with OPA and Kube-mgmt?
+
+The Kubernetes API Server is configured to query OPA for admission control decisions when objects (e.g., Pods, Services, etc.) are created, updated, or deleted. The API Server sends the entire Kubernetes object in the webhook request to OPA. OPA evaluates the policies it has loaded using the admission review as input. The diagram below shows the flow between a user making a request to the Kube-API server and how AdmissionReview and AdmissionRequests are made through OPA Gatekeeper.
```
Comment (shapirov103) [inline docs/addons/opa-gatekeeper.md:20]: "The diagram below" - was it missed?

File: docs/addons/opa-gatekeeper.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    rule: 'MustRunAs'
+    ranges:
+      # Forbid adding the root group.
+      - min: 1
+        max: 65535
+  readOnlyRootFilesystem: false
+```
+
+
+**Note: PSP's will be deprecated as of Kubernetes version 1.21 so please keep that in mind while you are evaluating this add on. To learn more please follow this [link](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/)**
```
Comment (shapirov103) [inline docs/addons/opa-gatekeeper.md:116]: minor: add on = add-on.

File: docs/addons/opa-gatekeeper.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    Image ID:      docker-pullable://openpolicyagent/opa@sha256:04ff8fce2afd1a3bc26260348e5b290e8d945b1fad4b4c16d22834c2f3a1814a
+    Port:          <none>
+    Host Port:     <none>
+    Args:
+      run
+      --server
+      --addr=localhost:8080
+```
+
+This is exactly what is defined in our example_allowed.yaml file so we know that our policy was deployed successfully using OPA Gatekeeper. For more information on OPA Gatekeeper policies, check out the GitHub repo which can be found [here](https://github.com/open-policy-agent/gatekeeper-library)
```
Comment (shapirov103) [inline docs/addons/opa-gatekeeper.md:195]: We need to show an example how customers will apply OPA policies with SSP. Where will those policies be located? Let's say we need to set cluster level or namespace level policies. The platform admin wants them to be in place when we provision clusters. While the examples are great, the practical aspects, e.g. an example solution is still missing. Should it be through the blueprint? GitOps? If so, how?

File: lib/addons/index.ts [hunk 7]
```diff
@@ -5,6 +5,7 @@ export { CalicoAddOn } from './calico'
 export { ContainerInsightsAddOn } from './container-insights'
 export { ClusterAutoScalerAddOn } from './cluster-autoscaler'
 export { ExternalDnsAddon } from './external-dns'
+export { OpaAddOn } from './OPA'
```
Comment (shapirov103) [inline lib/addons/index.ts:8]: OPA directory should be rename to opa lower-case (consistent naming convention). E.g. dns, nginx, etc. all lower case.

**Review by shapirov103 — COMMENTED** [review]
Looks good! Some comments.
We also need a PR for either SSP patterns or SSP workloads or both with an example.

File: docs/addons/opa-gatekeeper.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      # Forbid adding the root group.
+      - min: 1
+        max: 65535
+  readOnlyRootFilesystem: false
+```
+
+
+**Note: PSP's will be deprecated as of Kubernetes version 1.21 so please keep that in mind while you are evaluating this add on. To learn more please follow this [link](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/)**
+
+## Getting Started with OPA Gatekeeper
```
Comment (shapirov103) [inline docs/addons/opa-gatekeeper.md:118]: We need to call out any limitations. E.g. it looks like it is Linux only, things like that.  Will it work on Fargate? Let's add a section and call those out.

File: lib/addons/OPA/index.ts (hunk 1 above)
Comment (parkand1) [reply lib/addons/OPA/index.ts:3]: Addressed

File: lib/addons/index.ts (hunk 7 above)
Comment (parkand1) [reply lib/addons/index.ts:8]: addressed

File: lib/addons/OPA/index.ts (hunk 1 above)
Comment (shapirov103) [reply lib/addons/OPA/index.ts:3]: Is it? OPA directory is still upper case. OpaAddon no renamed?

Comment (parkand1) [reply lib/addons/OPA/index.ts:3]: changed name of directory to opa-gatekeeper and the addon is now named OpaGatekeeperAddOn - let me know if you'd like me to rename

File: docs/addons/opa-gatekeeper.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
+# What is OPA Gatekeeper?
+
+The Open Policy Agent (OPA, pronounced “oh-pa”) is an open source, general-purpose policy engine that unifies policy enforcement across the stack. OPA provides a high-level declarative language that lets you specify policy as code and simple APIs to offload policy decision-making from your software. You can use OPA to enforce policies in microservices, Kubernetes, CI/CD pipelines, API gateways, and more. OPA uses a policy language known as Rego which is a query language which was purpose built to support structured document models such as JSON. To learn more about Rego check out this [link](https://www.openpolicyagent.org/docs/latest/policy-language/).
```
Comment (kcoleman731) [inline docs/addons/opa-gatekeeper.md:3]: across what stack?

File: lib/addons/opa-gatekeeper/index.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+export class OpaGatekeeperAddOn implements ClusterAddOn {
+
+    readonly options?: OpaGatekeeperAddOnProps;
+
+    constructor(props?: OpaGatekeeperAddOnProps) {
+        this.options = { ...opagatekeeperAddonDefaults, ...props };
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+        const props = this.options;
```
Comment (kcoleman731) [inline lib/addons/opa-gatekeeper/index.ts:53]: Assume you want to use the props?

File: docs/addons/opa-gatekeeper.md (hunk 4 above)
Comment (parkand1) [reply docs/addons/opa-gatekeeper.md:20]: It shows up for me when I preview in vscode - let me know if you still can't see the diagram

File: docs/addons/opa-gatekeeper.md (hunk 8 above)
Comment (shapirov103) [reply docs/addons/opa-gatekeeper.md:110]: @parkand1 ^^

File: lib/addons/opa-gatekeeper/index.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+     /**
+     * Image pullPolicy
+     */
+    labelNamespacePullPolicy?: string,
+
+    /**
+     * Frequency with which audit is run
+     */
+    auditInterval?: string,
```
[the remark this answers is not in the document]
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:30]: addressed - changed to integer

File: docs/addons/opa-gatekeeper.md (hunk 8 above)
Comment (parkand1) [reply docs/addons/opa-gatekeeper.md:110]: addressed

File: lib/addons/opa-gatekeeper/index.ts (hunk 10 above)
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts]: addressed

File: lib/addons/opa-gatekeeper/index.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
+import { ClusterAddOn, ClusterInfo } from "../../spi"; 
+
+/**
+ * Properties available to configure with OPA Gatekeeper. The Helm chart automatically sets the Gatekeeper flag --exempt-namespace={{ .Release.Namespace }} in order to exempt the namespace where the chart is installed, and adds the admission.gatekeeper.sh/ignore label to the namespace during a post-install hook.
+ */
+ export interface OpaGatekeeperAddOnProps {
+    /**
+     * Add labels to the namespace during post install hooks
+     */
+    labelNamespaceEnabled?: string,
```
[the remark this answers is not in the document]
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:10]: I decided that taking this approach is too complicated and will confuse customers so took the same approach as the Calico add-on.

File: lib/addons/opa-gatekeeper/index.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * Frequency with which audit is run
+     */
+    auditInterval?: string,
+
+    /**
+     * Values to pass to the chart as per https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/README.md
+     */
+    values?: {
```
[the remark this answers is not in the document]
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:35]: I decided that taking this approach is too complicated and will confuse customers so took the same approach as the Calico add-on.

**Review by kcoleman731 — APPROVED** [review]

File: lib/addons/opa-gatekeeper/index.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
+import { ClusterAddOn, ClusterInfo } from "../../spi";
+
+/**
+ * Configuration options for the add-on.
+ */
+export interface OpaGatekeeperAddOnProps {
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:6]: You are no using any of the properties defined in the options. All of them must be used for provisioning.

**Review by shapirov103 — CHANGES_REQUESTED** [review]
Please see my comment on the OpaGatekeeperAddOnProps. Overall, getting there!

File: lib/addons/opa-gatekeeper/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (shapirov103) [reply lib/addons/opa-gatekeeper/index.ts]: We are using values across the board to allow for customizations, which most customers will want. Let's keep the values in the place to allow passing helm chart values.

File: lib/addons/opa-gatekeeper/index.ts (hunk 14 above)
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:6]: addressed

File: lib/addons/opa-gatekeeper/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts]: addressed

File: lib/addons/opa-gatekeeper/index.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        const props = this.options;
+
+        clusterInfo.cluster.addHelmChart("opagatekeeper-addon", {
+            chart: "gatekeeper",
+            release: "gatekeeper",
+            repository: "https://open-policy-agent.github.io/gatekeeper/charts",
+            version: props.version,
+            namespace: props.namespace,
+            values: {}
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:56]: You are not leveraging values that customers can pass to your addon. You can test by passing any values (e.g. label related) to the addon, they won't be applied since you set values to `{}`

Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts]: addressed

File: lib/addons/opa-gatekeeper/index.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    namespace?: string;
+
+    /**
+     * Helm chart version to use to install.
+     * @default 3.6.0-beta.3
+     */
+    version?: string;
+
+    /**
+     * Values for the Helm chart.
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:21]: Please add a reference to the chart values supported by helm that you mentioned in slack.

**Comment by shapirov103** [comment]
Let's resolve the merge (rebase) and I will kick off the tests.

File: lib/addons/opa-gatekeeper/index.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * 	Whether or not Helm should wait until all Pods, PVCs, Services, and minimum number of Pods of a Deployment, StatefulSet, or ReplicaSet are in a ready state before marking the release as successful.
+     * @default true
+     */
+    wait?: boolean;
+    /**
+     * Setting validatingWebhookFailurePolicy to True
+     * @default true 
+     */
+    disableValidatingWebhook?: boolean;
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:29]: This option is not leveraged for the helm chart.

File: lib/addons/opa-gatekeeper/index.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Helm chart version to use to install.
+     * @default 3.7.0-beta.1
+     */
+    version?: string;
+    /**
+     * 	Whether or not Helm should wait until all Pods, PVCs, Services, and minimum number of Pods of a Deployment, StatefulSet, or ReplicaSet are in a ready state before marking the release as successful.
+     * @default true
+     */
+    wait?: boolean;
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:24]: This option is not leveraged for the helm chart. That means that if customers pass this value, it will have no impact on the provisioned add-on.

File: lib/addons/opa-gatekeeper/index.ts (hunk 17 above)
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:34]: addressed

File: lib/addons/opa-gatekeeper/index.ts (hunk 18 above)
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts:28]: Removed - not necessary for the helm chart

File: lib/addons/opa-gatekeeper/index.ts [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+        const props = this.options;
+
+        clusterInfo.cluster.addHelmChart("opagatekeeper-addon", {
+            chart: "gatekeeper",
+            release: "gatekeeper",
+            repository: "https://open-policy-agent.github.io/gatekeeper/charts",
+            version: props.version,
+            namespace: props.namespace,
+            values: {
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:64]: You have declared the following options that customer can supply: values (since there a lot of customizations customers can apply) and disableValidatingWebhook. You are not using them in the helm chart. Please take a look at other add-ons that supply values and manipulate them. For example in my acm branch: https://github.com/aws-quickstart/ssp-amazon-eks/blob/feature/acm-integration/lib/addons/nginx/index.ts

Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts]: Addressed - let me know if this is the approach you were referring to

File: lib/addons/opa-gatekeeper/index.ts [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+
+    deploy(clusterInfo: ClusterInfo): void {
+
+        const props = this.options;
+
+        const values = { ...props.values ?? {}};
+
+        setPath(values, 'disableValidatingWebhook', false)
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:72]: You are hardcoding values for disableValidationWebhook and labelNamespace, basically means that *regardless* of what customer passes, you will set them to these values. This is wrong, you should propagate values that you allow customers to pass. 

Imagine customer is using your add-on like this:

```typescript
new OpaGateKeeperAddOn({
   labelNamespace: false,
   disableValidationWebhook: true
);
```

The above will have no effect.

File: package.json [hunk 21]
```diff
@@ -39,14 +39,17 @@
         "@aws-cdk/aws-iam": "1.119.0",
         "@aws-cdk/aws-logs": "1.119.0",
         "@aws-cdk/aws-secretsmanager": "1.119.0",
-        "@aws-cdk/core": "1.119.0",
+        "@aws-cdk/core": "^1.119.0",
+        "@aws-cdk/lambda-layer-kubectl": "^1.119.0",
```
Comment (shapirov103) [inline package.json:43]: Why do we need this dependency on lambdalayer?

File: package.json [hunk 22]
```diff
@@ -39,14 +39,17 @@
...
         "@aws-cdk/pipelines": "1.119.0",
         "@types/bcrypt": "^5.0.0",
         "aws-sdk": "^2.952.0",
         "bcrypt": "^5.0.1",
         "js-yaml": "3.13.1",
+        "object-utils": "^0.0.1-SNAPSHOT",
```
Comment (shapirov103) [inline package.json:49]: Please remove these dependencies, we don't need them.

**Review by shapirov103 — COMMENTED** [review]
There are many failed checks from github actions. Along with my comments, could you please take a look at those.

File: package.json (hunk 22 above)
Comment (parkand1) [reply package.json:49]: addressed

File: package.json (hunk 21 above)
Comment (parkand1) [reply package.json:43]: addressed

Comment (parkand1) [reply package.json:43]: Removed based on our conversation

File: package.json (hunk 22 above)
Comment (parkand1) [reply package.json:49]: addressed

File: package.json [hunk 23]
```diff
@@ -40,14 +40,15 @@
         "@aws-cdk/aws-iam": "1.119.0",
         "@aws-cdk/aws-logs": "1.119.0",
         "@aws-cdk/aws-secretsmanager": "1.119.0",
-        "@aws-cdk/core": "1.119.0",
+        "@aws-cdk/core": "^1.119.0",
```
Comment (shapirov103) [inline package.json:43]: why this change?

File: lib/addons/opa-gatekeeper/index.ts (hunk 17 above)
Comment (shapirov103) [reply lib/addons/opa-gatekeeper/index.ts:28]: how is it addressed? this value is not used in the add-on. you hardcode it to false.

File: lib/addons/opa-gatekeeper/index.ts [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+        this.options = { ...defaultProps, ...props };
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+
+        const props = this.options;
+
+        const values = { ...props.values ?? {}};
+
+        values['gatekeeper-configurations'] = {
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:71]: How did you come up with 'gatekeeper-configurations' and 'configurations' attributes? This will produce values like this:
```yaml
gatekeeper-configurations:
  configurations:
     disableValidatingWebhook: false
     labelNamespace: true
```

You need:
```yaml
disableValidatingWebhook: false
labelNamespace: true
```

File: package.json (hunk 23 above)
Comment (parkand1) [reply package.json]: removed the ^ and looks like that was breaking the CI

File: package.json [hunk 25]
```diff
@@ -48,6 +48,7 @@
         "js-yaml": "3.13.1",
         "sync-request": "6.1.0",
         "user": "^0.0.0",
+        "utils": "^0.3.1",
```
Comment (shapirov103) [inline package.json:51]: Why is it needed? Are you using it anywhere?

**Comment by shapirov103** [comment]
e2e faiiling. it is appmesh in this case, not really a descriptive error message, but it is related to the OPA add-on.

File: lib/addons/opa-gatekeeper/index.ts (hunk 24 above)
Comment (parkand1) [reply lib/addons/opa-gatekeeper/index.ts]: removed and took a different approach and as shown in my latest push

File: package.json (hunk 25 above)
Comment (parkand1) [reply package.json]: Removed

File: lib/addons/opa-gatekeeper/index.ts [hunk 26]
```diff
@@ -0,0 +1,57 @@
+import { ClusterAddOn, ClusterInfo } from "../../spi";
+// import { AppMeshAddOn } from '../appmesh/index';
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:2]: please remove this comment

File: lib/addons/opa-gatekeeper/index.ts [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+    deploy(clusterInfo: ClusterInfo): void {
+        clusterInfo.cluster.addHelmChart("OpaGatekeeperAddOn", {
+            chart: "gatekeeper",
+            release: "gatekeeper",
+            repository: "https://open-policy-agent.github.io/gatekeeper/charts",
+            version: this.options.version,
+            namespace: this.options.namespace,
+            values: this.options.values
+        });
+        // OpaGatekeeperAddOn.node.addDependency(AppMeshAddOn)
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:55]: Please remove this comment

File: lib/addons/opa-gatekeeper/index.ts [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+const defaultProps: OpaGatekeeperAddOnProps = {
+    namespace: 'gatekeeper-system',
+    version: '3.6.0',
+    values: {}
+};
+
+export class OpaGatekeeperAddOn implements ClusterAddOn {
+    private options: OpaGatekeeperAddOnProps;
+    static node: any;
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:41]: Why is it needed? especially static.

File: lib/addons/opa-gatekeeper/index.ts [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+    
+    constructor(props?: OpaGatekeeperAddOnProps) {
+        this.options = { ...defaultProps, ...props };
+    }
+    deploy(clusterInfo: ClusterInfo): void {
```
Comment (shapirov103) [inline lib/addons/opa-gatekeeper/index.ts:46]: new line before method

**Comment by shapirov103** [comment]
@parkand1  the e2e tests succeeded and I will merge the PR. Can you create a task for SSP-Patterns on how to use this add-on in practice? For example, create a blueprint with teams and clear OPA policies that can be demonstrable with the workloads repository.

**Comment by shapirov103** [comment]
Corresponding issue in SSP patterns:
https://github.com/aws-samples/ssp-eks-patterns/issues/4

**Review by shapirov103 — APPROVED** [review]
LGTM
